File: src/checks/runs.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

// Extended context type to store run ID across checks
type RunContext = CheckContext & { runId?: string; workflowId?: string };

export const runCreateCheck: Check = {
  name: 'Run Create',
  description: 'Create execution run',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.createdCustomerId || ctx.testCustomerId;

    if (!customerId) {
      return {
        name: 'Run Create',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      const result = await client.recordRun({
        customerId,
        workflow: 'health-check-workflow',
        events: [
          { eventType: 'task.start', description: 'Starting health check task' },
          { eventType: 'task.process', quantity: 1, units: 'tasks' },
          { eventType: 'task.complete', description: 'Health check completed' },
        ],
        status: 'COMPLETED',
      });

      const duration = performance.now() - start;

      // Store the run ID for subsequent checks
      (ctx as RunContext).runId = result.run.id;

      return {
        name: 'Run Create',
        success: true,
        duration,
        message: result.run.id,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;
      return {
        name: 'Run Create',
        success: false,
        duration,
        message: err.message || 'Failed to create run',
        details: err.code,
      };
    }
  },
};

export const runTimelineCheck: Check = {
  name: 'Run Timeline',
  description: 'Verify run events recorded',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.createdCustomerId || ctx.testCustomerId;

    if (!customerId) {
      return {
        name: 'Run Timeline',
        success: false,
        duration: 0,
        message: 'No customer ID available',
      };
    }

    try {
      // Create a run and then retrieve it
      const createResult = await client.recordRun({
        customerId,
        workflow: 'health-check-timeline',
        events: [
          { eventType: 'event1', description: 'First event' },
          { eventType: 'event2', description: 'Second event' },
          { eventType: 'event3', description: 'Third event' },
        ],
        status: 'COMPLETED',
      });

      const runResult = await client.getRunTimeline(createResult.run.id);
      const duration = performance.now() - start;

      const eventCount = runResult.timeline?.length || 0;

      return {
        name: 'Run Timeline',
        success: true,
        duration,
        message: `${eventCount} events recorded`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;
      return {
        name: 'Run Timeline',
        success: false,
        duration,
        message: err.message || 'Failed to get run timeline',
        details: err.code,
      };
    }
  },
};

export const runEndCheck: Check = {
  name: 'Run End',
  description: 'End run with status',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.createdCustomerId || ctx.testCustomerId;

    if (!customerId) {
      return {
        name: 'Run End',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      const sdk = (client as unknown as {
        sdk: {
          endRun?: (runId: string, data: { status: string; summary?: string }) => Promise<{ success: boolean }>;
          startRun?: (data: { customerId: string; workflow: string }) => Promise<{ id: string }>;
        };
      }).sdk;

      if (!sdk.endRun || !sdk.startRun) {
        const duration = performance.now() - start;
        return {
          name: 'Run End',
          success: true,
          duration,
          message: 'Skipped (endRun/startRun not available)',
          details: 'The endRun or startRun method is not available in the SDK',
        };
      }

      // First create a run to end
      const startResult = await sdk.startRun({
        customerId,
        workflow: 'health-check-end-test',
      });

      await sdk.endRun(startResult.id, {
        status: 'completed',
        summary: 'Health check completed successfully',
      });
      const duration = performance.now() - start;

      return {
        name: 'Run End',
        success: true,
        duration,
        message: `Run ${startResult.id} ended`,
        details: 'Status: completed',
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Run End',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Run End',
        success: false,
        duration,
        message: err.message || 'Failed to end run',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const emitEventCheck: Check = {
  name: 'Emit Event',
  description: 'Emit single event to run',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const runId = (ctx as RunContext).runId;

    if (!runId) {
      return {
        name: 'Emit Event',
        success: true,
        duration: 0,
        message: 'Skipped (no run ID available)',
        suggestion: 'Run Run Create check first',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          emitEvent?: (data: { runId: string; type: string; data: Record<string, unknown> }) => Promise<{ success: boolean }>;
        };
      }).sdk;

      if (!sdk.emitEvent) {
        const duration = performance.now() - start;
        return {
          name: 'Emit Event',
          success: true,
          duration,
          message: 'Skipped (emitEvent not available)',
          details: 'The emitEvent method is not available in the SDK',
        };
      }

      await sdk.emitEvent({
        runId,
        type: 'tool_call',
        data: { tool: 'health-check', input: { test: true } },
      });
      const duration = performance.now() - start;

      return {
        name: 'Emit Event',
        success: true,
        duration,
        message: 'Event emitted',
        details: `Run: ${runId}, Type: tool_call`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Emit Event',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Emit Event',
        success: false,
        duration,
        message: err.message || 'Failed to emit event',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const emitEventsBatchCheck: Check = {
  name: 'Emit Events Batch',
  description: 'Emit multiple events in one request',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const runId = (ctx as RunContext).runId;

    if (!runId) {
      return {
        name: 'Emit Events Batch',
        success: true,
        duration: 0,
        message: 'Skipped (no run ID available)',
        suggestion: 'Run Run Create check first',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          emitEventsBatch?: (events: Array<{ runId: string; type: string; data: Record<string, unknown> }>) => Promise<{ success: boolean; count: number }>;
        };
      }).sdk;

      if (!sdk.emitEventsBatch) {
        const duration = performance.now() - start;
        return {
          name: 'Emit Events Batch',
          success: true,
          duration,
          message: 'Skipped (emitEventsBatch not available)',
          details: 'The emitEventsBatch method is not available in the SDK',
        };
      }

      const result = await sdk.emitEventsBatch([
        { runId, type: 'llm_call', data: { model: 'gpt-4', tokens: 100 } },
        { runId, type: 'llm_call', data: { model: 'gpt-4', tokens: 200 } },
      ]);
      const duration = performance.now() - start;

      return {
        name: 'Emit Events Batch',
        success: true,
        duration,
        message: `${result.count || 2} events emitted`,
        details: `Run: ${runId}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Emit Events Batch',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Emit Events Batch',
        success: false,
        duration,
        message: err.message || 'Failed to emit events batch',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const recordRunCheck: Check = {
  name: 'Record Run',
  description: 'Record complete run in one call',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.createdCustomerId || ctx.testCustomerId;
    const workflowId = (ctx as RunContext).workflowId || 'health-check-record';

    if (!customerId) {
      return {
        name: 'Record Run',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      const result = await client.recordRun({
        customerId,
        workflow: workflowId,
        events: [
          { eventType: 'start', description: 'Run started' },
          { eventType: 'process', quantity: 5, units: 'items' },
          { eventType: 'complete', description: 'Run finished' },
        ],
        status: 'COMPLETED',
      });
      const duration = performance.now() - start;

      return {
        name: 'Record Run',
        success: true,
        duration,
        message: `Recorded: ${result.run.id}`,
        details: `Workflow: ${workflowId}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Record Run',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Record Run',
        success: false,
        duration,
        message: err.message || 'Failed to record run',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/runs.py
"""Execution run checks."""
import uuid
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _run_create_check(ctx: CheckContext) -> CheckResult:
    """Create a workflow run."""
    customer_id = ctx.created_customer_id or ctx.test_customer_id
    if not customer_id:
        return CheckResult(
            name="run_create",
            success=False,
            duration=0,
            message="No customer ID available"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        # Check if SDK supports runs
        if not hasattr(client, 'start_run') and not hasattr(client, 'record_run'):
            return CheckResult(
                name="run_create",
                success=True,
                duration=0,
                message="Run tracking not available in SDK",
                details="Skipping run tests"
            )

        workflow_slug = f"health-check-{uuid.uuid4().hex[:8]}"
        correlation_id = f"health_{uuid.uuid4().hex[:8]}"

        # Prefer record_run as it's simpler and handles workflow creation automatically
        if hasattr(client, 'record_run'):
            result = client.record_run(
                customer_id=customer_id,
                workflow=workflow_slug,  # SDK auto-creates workflow from slug
                status="COMPLETED",
                events=[
                    {
                        "eventType": "test.event",
                        "quantity": 100,
                        "units": "tokens",
                        "description": "Health check test event"
                    }
                ]
            )
            run_info = getattr(result, 'run', result)
            run_id = getattr(run_info, 'id', str(result))
            ctx.run_id = run_id

            return CheckResult(
                name="run_create",
                success=True,
                duration=0,
                message=f"Created and completed run {run_id}",
                details=f"workflow: {workflow_slug}"
            )

        # Fallback to start_run if record_run not available
        if hasattr(client, 'start_run'):
            # Must create workflow first for start_run
            if hasattr(client, 'create_workflow'):
                try:
                    workflow = client.create_workflow(
                        name="Health Check Workflow",
                        slug=workflow_slug,
                        product_surface="AGENT"
                    )
                    workflow_id = workflow.id
                except Exception:
                    # Workflow might already exist
                    workflow_id = workflow_slug
            else:
                return CheckResult(
                    name="run_create",
                    success=True,
                    duration=0,
                    message="Run tracking requires workflow creation",
                    details="create_workflow not available in SDK"
                )

            run = client.start_run(
                customer_id=customer_id,
                workflow_id=workflow_id,
                correlation_id=correlation_id
            )
            run_id = run.id

            # Emit an event if supported
            if hasattr(client, 'emit_event'):
                client.emit_event(
                    run_id=run_id,
                    event_type="test.event",
                    quantity=100,
                    units="tokens"
                )

            # End run
            if hasattr(client, 'end_run'):
                client.end_run(run_id, status="COMPLETED")

            ctx.run_id = run_id

            return CheckResult(
                name="run_create",
                success=True,
                duration=0,
                message=f"Created and completed run {run_id}",
                details=f"workflow: {workflow_slug}"
            )

        return CheckResult(
            name="run_create",
            success=True,
            duration=0,
            message="Run methods available but could not execute",
            details="Skipping run tests"
        )
    except Exception as e:
        return CheckResult(
            name="run_create",
            success=False,
            duration=0,
            message=f"Run creation failed: {e}"
        )


run_create_check = Check(
    name="run_create",
    description="Create and complete a workflow run",
    run=_run_create_check
)


async def _run_timeline_check(ctx: CheckContext) -> CheckResult:
    """Get run timeline."""
    if not ctx.run_id:
        return CheckResult(
            name="run_timeline",
            success=True,
            duration=0,
            message="No run ID available",
            details="Run create check may have been skipped"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'get_run_timeline'):
            return CheckResult(
                name="run_timeline",
                success=True,
                duration=0,
                message="Run timeline not available in SDK",
                details="Skipping timeline check"
            )

        timeline = client.get_run_timeline(ctx.run_id)

        # Handle different response formats
        if hasattr(timeline, 'events'):
            event_count = len(timeline.events)
        elif isinstance(timeline, list):
            event_count = len(timeline)
        else:
            event_count = 1

        return CheckResult(
            name="run_timeline",
            success=True,
            duration=0,
            message=f"Retrieved timeline with {event_count} events"
        )
    except Exception as e:
        return CheckResult(
            name="run_timeline",
            success=False,
            duration=0,
            message=f"Failed to get timeline: {e}"
        )


run_timeline_check = Check(
    name="run_timeline",
    description="Get run timeline",
    run=_run_timeline_check
)
